cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -485,6 +644,37 @@ ArrayKernelExec ArithmeticExecFromOp(detail::GetTypeId get_id) {
   }
 }
 
+// For kernels that always return floating results
+template <template <typename... Args> class KernelGenerator, typename Op>
+ArrayKernelExec IntToDoubleExecFromOp(detail::GetTypeId get_id) {
```

Comment:
*For context:* There are a variety of generator dispatchers in the compute layer and their names are inconsistent (this is indirectly related to ARROW-9161). There has been previous work in [renaming them for consistency](https://github.com/apache/arrow/pull/7461#issuecomment-645623392) but looking at the codebase, we will need another pass.

I suggest to change the name `IntToDoubleExecFromOp` to `GenerateArithmeticWithFloatOutType`.